apps/system/js/base_icon.js

```diff
@@ -0,0 +1,199 @@
+/* global Service, BaseUI */
+'use strict';
+
+(function(exports) {
+  /**
+   * BaseIcon is a base class for all icons on statusbar
+   * @param {Object} manager The manager object
+   *                         who has the information for the icon
+   * @param {Number} [index] Specify the index if it is not a singleton
+   */
+  var BaseIcon = function(manager, index) {
+    if (index !== undefined) {
+      this.index = index;
+      this.debug('index=',index);
+    } else {
+      this.index = '';
+    }
+    this.amendProperty();
+    this.manager = manager;
+    this.publish('created');
+  };
+  BaseIcon.prototype = Object.create(BaseUI.prototype);
+  BaseIcon.prototype.constructor = BaseIcon;
+  BaseIcon.prototype.DEBUG = false;
+  BaseIcon.prototype.camalToDash = function(strings) {
+    var i = 0;
+    var ch = '';
+    while (i <= strings.length) {
+      var character = strings.charAt(i);
+      if (character !== character.toLowerCase()) {
+        if (ch === '') {
+          ch += character.toLowerCase();
+        } else {
+          ch += '-' + character.toLowerCase();
+        }
+      } else {
+        ch += character;
+      }
+      i++;
+    }
+    return ch;
+  };
+  BaseIcon.prototype.amendProperty = function() {
+    if (!this.name) {
+      throw new Error('please specify a name when constructing a new icon');
+    }
+    var pureName = this.name.replace(/Icon$/, '');
+    var dashPureName = this.camalToDash(pureName);
+    this.instanceID = 'statusbar-' + dashPureName;
+    if (!this.CLASS_LIST) {
+      this.CLASS_LIST = 'sb-icon sb-icon-' + dashPureName;
+    }
+    if (!this.l10nId) {
+      this.l10nId = 'statusbar' + pureName;
+    }
+  };
+  BaseIcon.prototype.EVENT_PREFIX = 'icon';
+  BaseIcon.prototype.name = 'BaseIcon';
+  BaseIcon.prototype.containerElement = document.getElementById('statusbar');
+  // Overload me
+  BaseIcon.prototype.instanceID = 'statusbar-base';
+  BaseIcon.prototype.additionalProperties = '';
+  BaseIcon.prototype.role = 'listitem';
+  /**
+   * This flag is normally true because most icons want to update
+   * the status once it's rendered. But for certain special icon
+   * they don't want to be updated on started, set this flag to false.
+   * @type {Boolean}
+   */
+  BaseIcon.prototype.UPDATE_ON_START = true;
+  BaseIcon.prototype._fetchElements = function() {
+    this.element = document.getElementById(this.instanceID);
+  };
+  BaseIcon.prototype.show = function(force) {
+    if (!this.element) {
+      return;
+    }
+    var hidden = this.element.hidden;
```

Comment:
I wonder if we could keep the `isHidden` information in JS? Not sure if reading this property aggressively (since we will reach this line or the one in `hide()` every time `update()` is called) will cause any JS <--> C performance issue.